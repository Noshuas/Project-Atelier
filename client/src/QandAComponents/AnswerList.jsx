import React from 'react';
import Answer from './Answer.jsx';

const AnswerList = (props) => {
  if (!props.moreAnswers) {
    return (
      <div>
        {props.answers.map((answer, index) => {
          if (index < 2) {
            return <Answer info={answer} key={index}/>;
          }
        })}
        <ExandAnswers handleMoreAnswers={props.handleMoreAnswers} numAnswers={props.answers.length}/>
      </div>
    );
  }
  return (
    <div className='answer-list'>
      {props.answers.map((answer, index) => {
        if (index < props.answers.length) {
          return <Answer info={answer} key={index}/>;
        }
      })}
      <a className='more-answers-lnk' href='#' onClick={props.handleMoreAnswers}>Collapse answers</a>
    </div>
  );
};

function ExandAnswers(props) {
  if (props.numAnswers === 0) {
    return <span>This question currently has no answers</span>;
  }
  if (props.numAnswers > 2) {
    return <a className='more-answers-lnk' href='#' onClick={props.handleMoreAnswers}> See more answers</a>;
  }
  return null;
}

export default AnswerList;