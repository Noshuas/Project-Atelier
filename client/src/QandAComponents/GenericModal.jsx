import React from 'react';
import ReactDOM from 'react-dom';
// could make more generic

const modal_styles = {
  position: 'fixed',
  top: '50%',
  left: '50%',
  padding: '50px',
  backgroundColor: 'white',
  zIndex: 1000,
  transform: 'translate(-50%, -50%)',
};

const form_style = {
  'max-height': 'calc(100vh - 300px)',
  'overflow-y': 'auto',
  'overflow-x': 'hidden'
};

const overlay_styles = {
  position: 'fixed',
  top: 0,
  left: 0,
  right: 0,
  bottom: 0,
  backgroundColor: 'rgba(0, 0, 0, .7)',
  zIndex: 1000
};

function GenericModal ({open, children, onClose}) {
  if (!open) {
    return null;
  }
  return ReactDOM.createPortal(
    <div>
      <div style={overlay_styles}/>
      <div style={modal_styles}>
        <button className="exit-modal" onClick={onClose}>Exit</button>
        <div style={form_style}>
          {children}
        </div>
      </div>
      {/* <button onClick={onClose}>Submit</button> */}
    </div>,
    document.getElementById('portal')
  );
}

export default GenericModal;