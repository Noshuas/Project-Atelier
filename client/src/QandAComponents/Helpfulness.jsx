import React, {useState, useEffect} from 'react';
import QAapiCalls from './QandAAPIcalls.js';

function Helpfulness (props) {

  let [feedback, setFeedback] = useState(false);

  let handleClick = (event) => {
    event.preventDefault();
    QAapiCalls.postHelpfullnessFeedback(props.QorA, props.id, 'helpful')
      .then( (res) => {
        setFeedback(true);
      })
      .catch( (err) => { console.log('err:', err); });
  };
  if (!feedback) {

    return (
      <>
        <span> Helpful? <button className="link-style-btn" onClick={handleClick}>Yes</button> ({props.helpfulness})
        </span>
      </>
    );
  }
  return (
    <>
      <span> Helpful?  Yes ({props.helpfulness + 1})  </span>
    </>
  );
}

export default Helpfulness;