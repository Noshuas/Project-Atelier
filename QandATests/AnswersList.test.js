/**
 * @jest-environment jsdom
 */
import React, { useState } from 'react';
import renderer from 'react-test-renderer';
import Helpfulness from '../client/src/QandAComponents/Helpfulness.jsx';
import GenericModal from '../client/src/QandAComponents/GenericModal.jsx';
import ReactDOM from 'react-dom';
import Helpers from '../client/src/QandAComponents/helpers.js';
import AnswerList from '../client/src/QandAComponents/AnswerList.jsx';
import { render, screen } from '@testing-library/react';

// Answers List Rendering
// what asnwers list nees
// answers, moreAnswer boolean, handleMoreAnswers func

// only loads two to start
test('Answer list loads two answers at the start', () => {
  let testData = [
    {
      'answer_id': 8,
      'body': 'What a great question!',
      'date': '2018-01-04T00:00:00.000Z',
      'answerer_name': 'metslover',
      'helpfulness': 8,
      'photos': [],
    },
    {
      'answer_id': 9,
      'body': 'What a great question!',
      'date': '2018-01-04T00:00:00.000Z',
      'answerer_name': 'metslover',
      'helpfulness': 8,
      'photos': [],
    },
    {
      'answer_id': 10,
      'body': 'What a great question!',
      'date': '2018-01-04T00:00:00.000Z',
      'answerer_name': 'metslover',
      'helpfulness': 8,
      'photos': [],
    }];
  const component = renderer.create(
    <AnswerList answers={testData} moreAnswers={false} handleMoreAnswers={() => {}} />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
// see more to load more

// does not show see more if there are none