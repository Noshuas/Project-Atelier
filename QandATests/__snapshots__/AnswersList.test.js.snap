// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Answer list loads two answers at the start 1`] = `
<div>
  <div
    className="answer-section"
  >
    <span>
      What a great question!
    </span>
    <div
      className="answersToolbar"
    >
      <span>
        by 
        metslover
         
        January 4, 2018
         |
        <span>
           Helpful? 
          <a
            href="#"
            onClick={[Function]}
          >
            Yes
          </a>
           (
          8
          )
        </span>
         | 
        <a
          href="#"
          onClick={[Function]}
        >
          Report
        </a>
      </span>
    </div>
  </div>
  <div
    className="answer-section"
  >
    <span>
      What a great question!
    </span>
    <div
      className="answersToolbar"
    >
      <span>
        by 
        metslover
         
        January 4, 2018
         |
        <span>
           Helpful? 
          <a
            href="#"
            onClick={[Function]}
          >
            Yes
          </a>
           (
          8
          )
        </span>
         | 
        <a
          href="#"
          onClick={[Function]}
        >
          Report
        </a>
      </span>
    </div>
  </div>
  <a
    className="more-answers-lnk"
    href="#"
    onClick={[Function]}
  >
     See more answers
  </a>
</div>
`;
